### DISCUSSION TITLE
How to serialize JSON output at backend?

### DISCUSSION THREAD

**Discussion by vinerz** [ueberdosis/tiptap#241] [General Questions]
Hi there! This question is kinda off-topic, but as you guys are the prosemirror magicians, I'd like to give a try.

I want to save the editor JSON document in my database, but render and filter the XSS at my backend using nodejs. As `prosemirror-model` heavily relies on the browser DOM API, I can't use it there and I also couldn't find any packages at npm capable of doing this out of the box.

Do you guys have any tips for this situation?

Thanks!

**Comment by jjangga0214** [comment]
First, I've never used tiptap as I've just got interest in it. However, if it's true that I understood your question well, I can give you a general advice. 

Another simple strategy might be possible. Just save data as-is without post-processing to database. Modify the data on-demand when loading on frontend. Major SPA frameworks such as vue.js or server-side template engines support that feature (parsing data as text rather then html so < and > are to be `&lt`+`;`, `&gt`+`;`) natively.

**Comment by vinerz** [comment]
Hey @jjangga0214, thanks for the suggestion! Unfortunately, I must inject the data back to the frontend as `v-html` so I can't trust user input.

Just for the record, I got it to work by providing a simulated DOM API using JSDOM, if anyone is interested.

It would be very nice to have your opinions on this code as well.

```js
import { JSDOM } from 'jsdom';
import { schema } from 'prosemirror-schema-basic';
import { DOMSerializer } from 'prosemirror-model';

const mainSerializer = DOMSerializer.fromSchema(schema);

const getNewDocument = () => {
  const dom = new JSDOM('<!DOCTYPE html><div id="content"></div>');
  const document = dom.window.document;
  const renderer = document.querySelector('div');

  return { document, renderer };
};

export const prosemirrorToHTML = (nodeObject) => {
  const { document, renderer } = getNewDocument();
  const node = schema.nodeFromJSON(nodeObject);

  mainSerializer.serializeFragment(node, { document }, renderer);
  return document.getElementById('content').innerHTML;
};
```

**Comment by erickwilder** [comment]
@vinerz The the best feature (for me) of Prosemirror is that it is based on a schema that _you have full control_ of. If you get a JSON from the backend (or anywhere else) that doesn't comply with it, the document will not be rendered.

You don't have to use a `v-html` if you store the JSON in the database. Just provide the JSON back when instantiating the editor and Prosemirror will take care of generating the proper markup out of it (a view) in the browser. If something tampers the data, the schema of your editor will not be matched and _only what's defined on it_ is considered as *valid markup*.

**Comment by holtwick** [comment]
For my project [Collect](https://collect-app.com/) I'm also trying to use the JSON document outside of a browser environment. I implemented something similar to #66 to export Markdown. Everything works nice so far, but the problem I run into is, that WebPack bundles TipTap and Vue.js with my code I want to run in my browser less environment (which is [JSContext](https://developer.apple.com/documentation/javascriptcore/jscontext) in my case).

So my proposal would be to completely separate schema related methods from the rest of the TipTap `Node` implementations to allow to generate the schema separated from the UI logic without losing the nice plugin infrastructure. Maybe this could be achieved by subclassing? `NodeSchema` > `NodeLogic`?

**Comment by cimabe** [comment]
I solved this by using [prosemirror-to-html-js](https://www.npmjs.com/package/prosemirror-to-html-js)
basic usage:

```javascript
import { Renderer } from 'prosemirror-to-html-js';

const content = {
  type: 'doc',
  content: [
    {
      type: 'paragraph',
      content: [
        {
          type: 'text',
          text: 'Example Paragraph'
        }
      ]
    }
  ]
};
const renderer = new Renderer();

const html = renderer.render(content);

```


adding additional Marks and Nodes:
```javascript
import { Node, Mark } from 'prosemirror-to-html-js';

class Strike extends Mark {
  matching() {
    return this.mark.type === 'strike';
  }

  tag() {
    return 's';
  }
}

class Quote extends Node {
  matching() {
    return this.node.type === 'quote';
  }

  tag() {
    return 'blockquote';
  }
}

renderer.addMark(Strike);
renderer.addNode(Quote);
```

**Comment by philippkuehn** [comment]
In v2 there is an `generateHTML` method to convert JSON to HTML. This method expects all used extensions and doesn’t use any DOM methods and can therefore also be used on the server side.

**Reply by GitHubish** [reply]
@philippkuehn the problem I have is that this `generateHTML()` method can't actually render what is executed in the `addNodeView()` method ([doc](https://www.tiptap.dev/guide/custom-extensions#node-views-advanced)). Is there a way to do this? Could Tiptap propose some kind of tool to preview a content from a JSON and extensions (including the one using the `addNodeView()`)?

**Reply by hanspagel** [reply]
Not planned yet, follow #1221 for updates
